### PR TITLE
Add example for JWT/keycloak

### PR DESCRIPTION
**Pull request by polandll**
This is staged for the AUTH upgrade.

### CODE REVIEW

File: examples/stargate-jwt-auth/start_jwt.sh [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+docker-compose up -d cass-2
+(docker-compose logs -f cass-1 &) | grep -q "is now part of the cluster"
+# Bring up the 3rd C* node
+docker-compose up -d cass-3
+(docker-compose logs -f cass-1 &) | grep -q "is now part of the cluster"
+
+# this is where a cql script must be run to insert data and grant permissions
+docker run --rm -it -v /Users/lorina.poland/CLONES/stargate/docker-images/examples/stargate-jwt-auth/scripts:/scripts  \
+-v /Users/lorina.poland/CLONES/stargate/docker-images/examples/stargate-jwt-auth/cqlshrc:/.cassandra/cqlshrc  \
+--env CQLSH_HOST=host.docker.internal --env CQLSH_PORT=9045  nuvo/docker-cqlsh
```
Comment (dougwettlaufer) [inline examples/stargate-jwt-auth/start_jwt.sh:22]: Path's need updated.

File: examples/stargate-jwt-auth/start_jwt.sh [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+# Bring up keycloak for handling JWTs
+docker-compose up -d keycloak
+
```
Comment (dougwettlaufer) [inline examples/stargate-jwt-auth/start_jwt.sh:26]: Might be good to setup the user here

```shell
echo "* Request for authorization"
TOKEN=$(curl -s --data "username=admin&password=admin&grant_type=password&client_id=admin-cli" http://localhost:4444/auth/realms/master/protocol/openid-connect/token | jq -r '.access_token')

echo "\n"
echo " * user creation\n"
curl -L -X POST 'http://localhost:4444/auth/admin/realms/stargate/users' \
-H 'Content-Type: application/json' \
-H "Authorization: bearer $TOKEN" \
--data-raw '{
    "username": "testuser1",
    "email": "foo@bar.com",
    "enabled": true,
    "emailVerified": true,
    "attributes": {
        "userid": [
            "9876"
        ],
        "role": [
            "web_user"
        ]
    },
    "credentials": [
        {
            "type": "password",
            "value": "testuser1",
            "temporary": "false"
        }
    ]
}'
```

File: examples/stargate-jwt-auth/README.md [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+INSERT INTO store.shopping_cart (userid, item_count, last_update_timestamp) VALUES ('1234', 5, toTimeStamp(toDate(now())));
+```
+
+### Grant role permissions to the table
+
+```cql
+CREATE ROLE IF NOT EXISTS 'web_user' WITH PASSWORD = 'web_user' AND LOGIN = TRUE;
+GRANT MODIFY ON TABLE store.shopping_cart TO web_user;
+GRANT SELECT ON TABLE store.shopping_cart TO web_user;
+```
```
Comment (dougwettlaufer) [inline examples/stargate-jwt-auth/README.md:43]: Not sure where you want to put it but this is a smoke test for JWT auth via REST

```
USER_TOKEN=$(curl -s --data "username=testuser1&password=testuser1&grant_type=password&client_id=user-service" http://localhost:4444/auth/realms/stargate/protocol/openid-connect/token | jq -r '.access_token')


curl -sL 'localhost:8082/v1/keyspaces/system/tables/local/rows/local' \
-H "X-Cassandra-Token: $USER_TOKEN" | jq .


curl -sL 'localhost:8082/v1/keyspaces/store/tables/shopping_cart/rows/9876' \
-H "X-Cassandra-Token: $USER_TOKEN" | jq .
{
  "count": 1,
  "rows": [
    {
      "item_count": 2,
      "userid": "9876",
      "last_update_timestamp": "2020-11-06T00:00:00Z"
    }
  ]
}

curl -sL 'localhost:8082/v1/keyspaces/store/tables/shopping_cart/rows/1234' \
-H "X-Cassandra-Token: $USER_TOKEN" | jq .
{
  "description": "Role unauthorized for operation: Not allowed to access this resource",
  "code": 401
}
```

File: examples/stargate-jwt-auth/docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+      - backend
+    ports:
+      - 8080:8080
+      - 8081:8081
+      - 8082:8082
+      - 8084:8084
+      - 9045:9042
+    mem_limit: 2G
+    environment:
+      - JAVA_OPTS=-XX:+CrashOnOutOfMemoryError -Xmx750M -Xms64M -Dstargate.auth_id=AuthJwtService -Dstargate.auth.jwt_provider_url=http://keycloak:4444/auth/realms/stargate/protocol/openid-connect/certs
```
Comment (dougwettlaufer) [inline examples/stargate-jwt-auth/docker-compose.yml:55]: If you want to use the JWT with cql you'll also need `-Dstargate.cql_use_auth_service=true`

File: examples/stargate-jwt-auth/docker-compose.yml [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    container_name: stargate-jwt
+    depends_on: 
+      - cass-1
+    networks:
+      - backend
+    ports:
+      - 8080:8080
+      - 8081:8081
+      - 8082:8082
+      - 18084:8084
```
Comment (dougwettlaufer) [inline examples/stargate-jwt-auth/docker-compose.yml:53]:
```suggestion
      - 8084:8084
```

File: examples/stargate-jwt-auth/docker-compose.yml (hunk 4 above)
Comment (polandll) [reply examples/stargate-jwt-auth/docker-compose.yml:55]: fixed

File: examples/stargate-jwt-auth/start_jwt.sh (hunk 1 above)
Comment (polandll) [reply examples/stargate-jwt-auth/start_jwt.sh]: This is outdated. I added a variable to use for the path to the git clone directory in the script. The script unfortunately requires an absolute path, and cannot use a relative path.

File: examples/stargate-jwt-auth/start_jwt.sh (hunk 2 above)
Comment (polandll) [reply examples/stargate-jwt-auth/start_jwt.sh:33]: I added this code to the startup script and made a note on the docs page that this step is completed in the startup script. I still show the step in the docs for the explanation.

File: examples/stargate-jwt-auth/README.md (hunk 3 above)
Comment (polandll) [reply examples/stargate-jwt-auth/README.md:43]: These tests are shown in the docs. I guess I can either add as a script in this repo, or just add a link to the docs in the README.

**Review by dougwettlaufer — DISMISSED** [review]

**Review by dougwettlaufer — APPROVED** [review]

**Pull request merged**